File: host-rs/src/vdd.rs
//! MikeTheTech Virtual Display Driver control, mirroring the C# `VddController`.

use crate::adb::run_process;
use crate::display_config::{self, MonitorInfo};
use crate::logline;
use windows::core::PCWSTR;
use windows::Win32::Graphics::Gdi::{ChangeDisplaySettingsExW, EnumDisplayDevicesW, DISPLAY_DEVICEW};

const PNPUTIL_TIMEOUT_MS: u64 = 8_000;
const POWERSHELL_TIMEOUT_MS: u64 = 15_000;

pub struct VddController {
    friendly_name: String,
    instance_id: String,
}

impl VddController {
    pub fn new() -> Self {
        Self {
            friendly_name: "Virtual Display Driver".to_string(),
            instance_id: r"ROOT\DISPLAY\0000".to_string(),
        }
    }

    fn run_pnputil(&self, arg: &str) -> bool {
        let r = run_process("pnputil.exe", &[arg, &self.instance_id], PNPUTIL_TIMEOUT_MS);
        if r.timed_out {
            logline!("[vdd] pnputil timeout ({arg})");
            return false;
        }
        r.code == 0 || r.code == 3010
    }

    fn run_powershell(&self, command: &str) -> (i32, String, String) {
        logline!("[vdd] {command}");
        let r = run_process(
            "powershell.exe",
            &["-NoProfile", "-NonInteractive", "-Command", command],
            POWERSHELL_TIMEOUT_MS,
        );
        (r.code, r.stdout, r.stderr)
    }

    pub fn enable(&self) {
        if self.run_pnputil("/enable-device") {
            logline!("[vdd] VDD enabled (fast pnputil).");
            return;
        }
        let cmd = format!(
            "$d = Get-PnpDevice -FriendlyName '{}' -ErrorAction SilentlyContinue; if ($d) {{ Enable-PnpDevice -InstanceId $d.InstanceId -Confirm:$false; 'OK' }} else {{ 'NOTFOUND' }}",
            self.friendly_name
        );
        let (code, stdout, stderr) = self.run_powershell(&cmd);
        if code != 0 && stderr.to_ascii_lowercase().contains("access is denied") {
            logline!("[vdd] Enable failed — needs administrator.");
        } else if stdout.contains("NOTFOUND") {
            logline!("[vdd] Enable: device not found");
        }
    }

    pub fn disable(&self) {
        self.remove_vdd_monitors();
        if self.run_pnputil("/disable-device") {
            logline!("[vdd] VDD disabled (fast pnputil).");
            return;
        }
        let cmd = format!(
            "$d = Get-PnpDevice -FriendlyName '{}' -ErrorAction SilentlyContinue; if ($d) {{ Disable-PnpDevice -InstanceId $d.InstanceId -Confirm:$false; 'OK' }} else {{ 'NOTFOUND' }}",
            self.friendly_name
        );
        let (code, stdout, stderr) = self.run_powershell(&cmd);
        if code != 0 && stderr.to_ascii_lowercase().contains("access is denied") {
            logline!("[vdd] Disable failed — needs administrator.");
        } else if stdout.contains("NOTFOUND") {
            logline!("[vdd] Disable: device not found");
        }
    }

    /// Poll until a monitor that wasn't in `before` appears (the VDD), or timeout.
    pub fn wait_for_monitor(&self, before: &[MonitorInfo], timeout_ms: u64) -> Option<MonitorInfo> {
        use std::time::{Duration, Instant};
        let start = Instant::now();
        let before_devices: Vec<&str> = before.iter().map(|m| m.device.as_str()).collect();
        loop {
            let current = display_config::get_monitors();
            let mut news: Vec<&MonitorInfo> =
                current.iter().filter(|m| !before_devices.contains(&m.device.as_str())).collect();
            if news.len() >= 1 {
                let chosen = choose_new_monitor(&mut news);
                return Some(chosen.clone());
            }
            if start.elapsed() > Duration::from_millis(timeout_ms) {
                // A leftover VDD phantom may already be present — reuse it.
                if let Some(f) = display_config::find_vdd_monitor(&current) {
                    logline!("[vdd] VDD phantom already present: {} {}x{} — reusing", f.device, f.width, f.height);
                    return Some(f);
                }
                logline!("[vdd] WaitForMonitor timeout");
                return None;
            }
            std::thread::sleep(Duration::from_millis(250));
        }
    }

    /// Names of the VDD's display devices (`\\.\DISPLAYn`) that the desktop currently sees.
    pub fn active_monitors(&self) -> Vec<String> {
        let mut names: Vec<String> = Vec::new();
        unsafe {
            let mut i = 0u32;
            loop {
                let mut dd = DISPLAY_DEVICEW::default();
                dd.cb = std::mem::size_of::<DISPLAY_DEVICEW>() as u32;
                if !EnumDisplayDevicesW(PCWSTR::null(), i, &mut dd, 0).as_bool() {
                    break;
                }
                let dev_name = wide(&dd.DeviceName);
                let mut mon = DISPLAY_DEVICEW::default();
                mon.cb = std::mem::size_of::<DISPLAY_DEVICEW>() as u32;
                if EnumDisplayDevicesW(PCWSTR(dd.DeviceName.as_ptr()), 0, &mut mon, 0).as_bool() {
                    let id = wide(&mon.DeviceID).to_ascii_uppercase();
                    if id.contains("MTT") {
                        names.push(dev_name);
                    }
                }
                i += 1;
            }
        }
        names
    }

    fn remove_vdd_monitors(&self) {
        let names = self.active_monitors();
        for n in &names {
            logline!("[vdd] Removing phantom monitor: {n}");
            let w: Vec<u16> = n.encode_utf16().chain(std::iter::once(0)).collect();
            const CDS_DETACH: u32 = 0x08;
            unsafe {
                let _ = ChangeDisplaySettingsExW(
                    PCWSTR(w.as_ptr()),
                    None,
                    None,
                    windows::Win32::Graphics::Gdi::CDS_TYPE(CDS_DETACH),
                    None,
                );
            }
        }
        if !names.is_empty() {
            std::thread::sleep(std::time::Duration::from_millis(100));
        }
    }
}

fn wide(buf: &[u16]) -> String {
    let end = buf.iter().position(|&c| c == 0).unwrap_or(buf.len());
    String::from_utf16_lossy(&buf[..end])
}

/// Among new monitors prefer a ~3:2 aspect (the tablet), else the first.
fn choose_new_monitor<'a>(news: &mut Vec<&'a MonitorInfo>) -> &'a MonitorInfo {
    news.sort_by(|a, b| a.device.cmp(&b.device));
    if let Some(pref) = news.iter().find(|m| {
        let ar = m.width as f64 / m.height as f64;
        (1.3..=1.55).contains(&ar)
    }) {
        return pref;
    }
    news[0]
}

File: host-zig/src/vdd.zig
const std = @import("std");
const w = @import("win.zig");

var vdd_active = false;

fn containsIgnoreCase(hay: []const u8, needle: []const u8) bool {
    if (needle.len == 0 or hay.len < needle.len) return false;
    var i: usize = 0;
    while (i + needle.len <= hay.len) : (i += 1) {
        var ok = true;
        var j: usize = 0;
        while (j < needle.len) : (j += 1) {
            if (std.ascii.toLower(hay[i + j]) != std.ascii.toLower(needle[j])) {
                ok = false;
                break;
            }
        }
        if (ok) return true;
    }
    return false;
}

/// Remove MTT phantom monitors so disabling the driver drops them from the topology.
fn detachPhantoms() void {
    var dd: w.DisplayDeviceA = undefined;
    var i: u32 = 0;
    while (i < 16) : (i += 1) {
        dd = std.mem.zeroes(w.DisplayDeviceA);
        dd.cb = @sizeOf(w.DisplayDeviceA);
        if (w.EnumDisplayDevicesA(null, i, &dd, 0) == 0) break;
        var mon: w.DisplayDeviceA = undefined;
        mon = std.mem.zeroes(w.DisplayDeviceA);
        mon.cb = @sizeOf(w.DisplayDeviceA);
        const name_z = nameToZ(&dd) orelse continue;
        if (w.EnumDisplayDevicesA(name_z, 0, &mon, 0) == 0) continue;
        const id = sliceZ(&mon.device_id);
        if (containsIgnoreCase(id, "MTT")) {
            w.logz("Removing phantom monitor: {s}\n", .{sliceZ(&dd.device_name)});
            _ = w.ChangeDisplaySettingsExA(name_z, null, null, 0x00000008, null);
        }
    }
    w.Sleep(100);
}

fn sliceZ(buf: []const u8) []const u8 {
    const n = std.mem.indexOfScalar(u8, buf, 0) orelse buf.len;
    return buf[0..n];
}

/// device_name as a NUL-terminated stack buffer (EnumDisplayDevices needs LPCSTR).
threadlocal var name_storage: [33]u8 = undefined;
fn nameToZ(dd: *w.DisplayDeviceA) ?[*:0]const u8 {
    const s = sliceZ(&dd.device_name);
    if (s.len == 0 or s.len >= name_storage.len) return null;
    @memcpy(name_storage[0..s.len], s);
    name_storage[s.len] = 0;
    return name_storage[0..s.len :0];
}

pub fn enable() void {
    if (vdd_active) return;
    w.logz("Enabling Virtual Display Driver (ROOT\\DISPLAY\\0000)...\n", .{});
    w.runHidden("pnputil.exe /enable-device \"ROOT\\DISPLAY\\0000\"");
    w.runHidden("powershell.exe -NoProfile -NonInteractive -Command \"$d = Get-PnpDevice -FriendlyName 'Virtual Display Driver' -ErrorAction SilentlyContinue; if ($d -and $d.Status -ne 'OK') { Enable-PnpDevice -InstanceId $d.InstanceId -Confirm:$false }\"");
    w.Sleep(500);
    // SDC_TOPOLOGY_SUPPLIED (0x80) | SDC_ALLOW_CHANGES (0x4): let the display
    // manager settle with the freshly enabled device, no path garbage.
    _ = w.SetDisplayConfig(0, null, 0, null, 0x00000080 | 0x00000004);
    _ = w.ChangeDisplaySettingsExA(null, null, null, 0, null);
    vdd_active = true;
    w.logz("Virtual Display Driver enabled successfully.\n", .{});
}

pub fn disable() void {
    w.logz("Disabling Virtual Display Driver (ROOT\\DISPLAY\\0000)...\n", .{});
    detachPhantoms();
    w.runHidden("pnputil.exe /disable-device \"ROOT\\DISPLAY\\0000\"");
    w.runHidden("powershell.exe -NoProfile -NonInteractive -Command \"$d = Get-PnpDevice -FriendlyName 'Virtual Display Driver' -ErrorAction SilentlyContinue; if ($d -and $d.Status -eq 'OK') { Disable-PnpDevice -InstanceId $d.InstanceId -Confirm:$false }\"");
    w.Sleep(300);
    _ = w.ChangeDisplaySettingsExA(null, null, null, 0, null);
    vdd_active = false;
    w.logz("Virtual Display Driver disabled successfully.\n", .{});
}
